app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java

```diff
@@ -535,4 +465,156 @@ public int setSelectionExpandWholeLines() {
+                _maxNumberDigits = 1;
+            } else if (number < 100) {
+                _maxNumberDigits = 2;
+            } else if (number < 1000) {
+                _maxNumberDigits = 3;
+            } else if (number < 10000) {
+                _maxNumberDigits = 4;
+            } else {
+                _maxNumberDigits = 5;
+            }
+            return _maxNumberDigits;
+        }
+
+        public boolean isVisibleAreaChanged() {
+            if (_visibleRect.top == _lastVisibleRect.top && _visibleRect.bottom == _lastVisibleRect.bottom) {
+                return false;
+            } else {
+                _lastVisibleRect.top = _visibleRect.top;
+                _lastVisibleRect.bottom = _visibleRect.bottom;
+                return true;
+            }
+        }
+
+        /**
+         * Draw line numbers.
+         *
+         * @param canvas the canvas on which the line numbers will be drawn.
+         */
+        public void draw(final Canvas canvas) {
+            if (!_editor.getLocalVisibleRect(_visibleRect)) {
+                return;
+            }
+
+            final CharSequence text = _editor.getText();
+            final Layout layout = _editor.getLayout();
+            if (text == null || layout == null) {
+                return;
+            }
+
+            final int offsetY = _editor.getPaddingTop();
+            final int layoutLineCount = layout.getLineCount();
+
+            // Only if the visible area or the layout line count changed
+            // We make a single pass through the text to determine
+            // 1. y positions and line numbers we want to draw
+            // 2. max line number (to gauge gutter width)
+            if (isVisibleAreaChanged() || layoutLineCount != _lastLayoutLineCount) {
+                _lastLayoutLineCount = layoutLineCount;
+                _maxNumber = 1;
+                _startNumber = 1;
+                _numberYPositions.clear();
+                _top = _visibleRect.top - _visibleRect.height();
+                _bottom = _visibleRect.bottom + _visibleRect.height();
+
+                for (int i = 0; i < layoutLineCount; i++) {
+                    System.err.println(i);
+                    final int start = layout.getLineStart(i);
+                    if (start == 0 || text.charAt(start - 1) == '\n') {
+                        final int y = layout.getLineBaseline(i);
+                        if (y < _top) {
+                            _startNumber++;
+                        } else if (y < _bottom) {
+                            _numberYPositions.add(y);
+                        }
+                        _maxNumber++;
+                    }
+                }
+            }
+
+            // Only if the text size or the max line number of digits changed
+            // Update the gutter parameters and set padding left
+            if (_paint.getTextSize() != _lastTextSize || getNumberDigits(_maxNumber) != _lastMaxNumberDigits) {
```

Comment:
So in the version I wrote I just compare the newly computed padding to the old padding.

I don't see how checking text size or max digits is more comprehensive. If either of these changes then the computed padding will change and that is sufficient to trigger an update.

IMO we are doing extra work here which is not necessary. (But please correct me if I am wrong)

Reply:
The padding wont automatically resize when the text size or the max line numbers of digits changed, so if just compare the newly computed padding with the old padding, this needs to update several parameters about the padding (especially the calculation of text width) every time, I think this is very inefficient, we know onDraw() be called very frequently.

Reply:
Same observation about not seeing a performance impact in real life.